MySQL: Track session-specific system variables changes

wazir-ahmed opened the pull request:
- Configure `session_track_system_variables` and `session_track_state_change` in all backend connections.
- Utilize notifications from backend servers to capture system variable changes that cannot be handled by `MySQL_Set_Stmt_Parser`
- Update both client and server variable maps based on backend responses.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Configurable session variable tracking: DISABLED (default), OPTIONAL, ENFORCED with runtime toggle and fast-forward support.
  * Proxy now respects per-server backoff when selecting backend connections and preserves client capability advertising when ENFORCED.
  * Backend-driven session-variable/state setup and extraction integrated into connection flow.

* **Tests**
  * New TAP tests covering OPTIONAL/ENFORCED modes, fast-forward paths, and MySQL-version specific behavior.
  * New CI infra and helper scripts for MySQL 5.6 single-instance testing.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wazir-ahmed committed:
MySQL: Track session-specific system variables changes

- Add new mysql variable `mysql-session_track_variables`.
- Configure `session_track_system_variables` and `session_track_state_change`
  on backend connections if the mysql variable is enabled.
- Utilize notifications from backend servers to capture system variable changes
  that cannot be handled by `MySQL_Set_Stmt_Parser`
- Update both client and server variable maps based on backend responses.
- TAP test to verify this patch.

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

renecannao committed:
docs: Add comprehensive Doxygen documentation for session variable tracking

Add detailed architectural documentation for PR 5166 session variable tracking:

- High-level architecture overview in session_track_variables struct explaining the 3-phase workflow
- Detailed documentation for handler_rc0_Process_Variables explaining the core processing workflow
- Technical implementation details for MySQL_Connection::get_variables protocol interface
- Configuration logic documentation for handler_again___verify_backend_session_track_variables
- Added inline comments explaining why session tracking is needed and performance considerations

This documentation provides a complete understanding of how MySQL session variable tracking integrates with ProxySQL's existing state machine and leverages MySQL's native session tracking capabilities.

renecannao commented:
# PR 5166: Comprehensive Review and Concerns - Detailed Analysis

First, excellent work on implementing session variable tracking! This is a valuable enhancement that addresses real limitations in ProxySQL's session state management. The architectural approach leveraging MySQL's native session tracking capabilities is sound and well-designed.

However, I have identified several critical issues and areas for improvement that should be addressed before merge. This analysis provides detailed technical background for each concern.

## 🔴 Critical Issues (Blockers)

### 1. Missing Capability Detection - Protocol Level Compatibility Gap

**Problem**: The implementation lacks proper capability detection for `CLIENT_SESSION_TRACK`, unlike the existing GTID tracking.

**Technical Background**:
In MySQL's client/server protocol, both client and server exchange capability flags during connection handshake. `CLIENT_SESSION_TRACK` is a capability flag that indicates the server supports MySQL 5.7+ session tracking features. When this flag is not set:

- Server does not track session variable changes
- Server cannot send session tracking information in OK packets
- Attempting to `SET session_track_system_variables` will likely fail
- MySQL client library functions like `mysql_session_track_get_first()` will not work

**Current Implementation Gap**:
```cpp
// ✅ GTID tracking correctly checks capability:
bool MySQL_Session::handler_again___verify_backend_session_track_gtids() {
    if ((mybe->server_myds->myconn->mysql->server_capabilities & CLIENT_SESSION_TRACK) == 0) {
        // the backend doesn't support CLIENT_SESSION_TRACK
        return ret; // Gracefully handles unsupported backends
    }
    // ... proceed with configuration
}

// ❌ Variable tracking attempts SET commands without capability check:
bool MySQL_Session::handler_again___verify_backend_session_track_variables() {
    if (mysql_thread___session_track_variables == session_track_variables::DISABLED) {
        return false;
    }
    // ❌ No CLIENT_SESSION_TRACK check - will try to SET on unsupported servers!
    if (mybe->server_myds->myconn->options.session_track_variables_sent == false) {
        // This will fail on MySQL < 5.7, MariaDB, SQLite3, etc.
        NEXT_IMMEDIATE_NEW(SETTING_SESSION_TRACK_VARIABLES);
    }
}
```

**Impact**: This will cause connection failures on:
- **MySQL < 5.7**: Session tracking was introduced in MySQL 5.7.4
- **MariaDB < 10.2**: MariaDB implemented this later than MySQL
- **ProxySQL's own admin interface**: Uses SQLite3 backend which doesn't support MySQL protocol extensions
- **sqlite3-server module**: SQLite3 backend in ProxySQL
- **Any non-MySQL backend that uses the MySQL Protocol**

**Fix Required**: Add the same capability check that GTID tracking uses before attempting any SET commands.

### 2. CLIENT_DEPRECATE_EOF Dependency Not Handled

**Technical Background**:
MySQL session tracking requires two capability flags to work properly:

1. **CLIENT_SESSION_TRACK** : Enables session tracking protocol support
2. **CLIENT_DEPRECATE_EOF** : Replaces EOF packets with OK packets, required for session tracking data

When `CLIENT_DEPRECATE_EOF` is not set, MySQL server sends EOF packets instead of OK packets with session tracking data. Session tracking information is only transmitted in OK packets, not EOF packets.

**Current State**: The implementation assumes backends support session tracking without verifying both required capabilities.

**Impact**: Even if a server supports `CLIENT_SESSION_TRACK`, without `CLIENT_DEPRECATE_EOF` the session tracking data cannot be transmitted to the client. I actually expect that `CLIENT_DEPRECATE_EOF` is a requirement for `CLIENT_SESSION_TRACK`.

## 🟡 Major Concerns

### 3. TAP Test Modifications Need Explanation

**Observation**: Two TAP tests were modified to explicitly disable session tracking:

```cpp
// test_binlog_reader-t.cpp:
MYSQL_QUERY_T(proxysql_admin, "SET mysql-session_track_variables=0");
MYSQL_QUERY_T(proxysql_admin, "LOAD MYSQL VARIABLES TO RUNTIME");

// reg_test_4264-commit_rollback-t.cpp:
MYSQL_QUERY_T(admin, "SET mysql-session_track_variables=0");
```

**Possible Root Causes**:
- **Test Failures**: Tests likely fail when session tracking is enabled due to the capability issues above
- **State Pollution**: The SET commands might affect test state in unexpected ways
- **Variable Behavior Changes**: Session tracking changes how variables are managed, potentially breaking test assumptions

**Action Required**: We need to understand the exact failure modes before merge. These test modifications suggest issues were encountered but worked around them instead of fixing the root cause.

### 4. Fast Forward Mode Compatibility

**Technical Background**:
Fast forward mode in ProxySQL provides high-performance query forwarding with minimal protocol overhead. It's designed for scenarios where ProxySQL acts as a transparent proxy with minimal processing.

**Session Tracking Impact**:
- **Packet Format Changes**: Session tracking adds data to OK packets, changing the protocol stream
- **Client Capability Mismatch**: Fast forward may assign connections to clients that don't support session tracking

**Current Gap**: No checking of client capabilities when assigning connections in fast forward mode.

**Impact**: Clients without session tracking support may receive packets they cannot parse or unexpected protocol overhead.

## 🔵 Enhancement Opportunities

### 5. Enhanced Configuration Modes (Production-Ready Design)

**Current Limitation**: Binary enable/disable (0/1) is too simplistic for production environments with mixed backends.

**Proposed Three-Tier Configuration**:

```cpp
enum mode {
    DISABLED = 0,  // Current behavior: feature completely disabled
    OPTIONAL = 1,  // Graceful fallback with logging for mixed environments
    ENFORCED = 2   // Strict mode: fail if requirements cannot be met
};
```

**Detailed Mode Behavior**:

**OPTIONAL Mode (1)**:
- Attempts to enable session tracking on supported backends
- Logs INFO message when backend doesn't support session tracking
- Logs WARNING if `mysql-enable_server_deprecate_eof` is false (required capability)
- Continues gracefully without session tracking on unsupported backends
- **Use Case**: Gradual rollout in mixed-version environments

**ENFORCED Mode (2)**:
- Automatically enables `CLIENT_DEPRECATE_EOF` regardless of `mysql-enable_server_deprecate_eof` setting
- Fails connection if `session_track_system_variables` or `session_track_state_change` cannot be set
- Logs ERROR and drops backend connection if requirements not met
- **Use Case**: Production environments requiring consistent behavior

### 6. Frontend Variable Propagation (Complete Implementation)

**Current Gap**: The implementation tracks variables internally for ProxySQL's state management but doesn't propagate changes to clients that support session tracking.

**Technical Requirements**:

1. **Client Capability Detection**: Check if client connection has `CLIENT_SESSION_TRACK` flag
2. **OK Packet Enhancement**: Add session tracking data to OK packets sent to supporting clients

**Implementation Complexity**:
- OK packet format changes require careful protocol handling
- Need to buffer session tracking data from backend and forward to client
- Must handle cases where client and backend support different levels of tracking
- `session_track_system_variables` or `session_track_state_change` themselves should become tracked variables

## 📋 Required Testing

### 7. Comprehensive CI Testing with Feature Enabled

**Current State**: Tests run with feature disabled (see TAP test modifications above)

**Required Testing Matrix**:

**A. Version Compatibility with mysql-session_track_variables=1 and mysql-session_track_variables=2**:
- MySQL 5.6 (no session tracking support)
- MySQL 8.0+ (full session tracking support)

**B. Protocol-Level Testing**:
- OK packet format validation
- Client compatibility verification
- Error handling for unsupported backends

## 🎯 Recommended Implementation Order

### Phase 1 - Critical Fixes (Merge Blockers)
1. **Add `CLIENT_SESSION_TRACKING` capability detection** before attempting SET commands
2. **Add `CLIENT_DEPRECATE_EOF` dependency handling**
3. **Fix TAP test failures** - understand root causes instead of disabling feature
4. **Add comprehensive error handling** for unsupported backends

### Phase 2 - Compatibility Hardening
1. **Fast forward mode compatibility checks**
2. **Comprehensive CI testing** with feature enabled
3. **Performance baseline measurement**

### Phase 3 - Production Enhancements
1. **Implement 0/1/2 configuration modes**
2. **Frontend variable propagation**
3. **Performance optimization** for high-frequency workloads
4. **Add metrics and monitoring** for session tracking usage

## 📝 Additional Technical Implementation Details

### Error Handling Strategy Example:
```cpp
bool MySQL_Session::handler_again___verify_backend_session_track_variables() {
    if (mysql_thread___session_track_variables == session_track_variables::DISABLED) {
        return false;
    }

    // Check if backend supports session tracking
    if ((mybe->server_myds->myconn->mysql->server_capabilities & CLIENT_SESSION_TRACK) == 0) {
        if (mysql_thread___session_track_variables == ENFORCED) {
            proxy_error("Backend %s:%d does not support required session tracking capabilities (CLIENT_SESSION_TRACK)",
                       mybe->host, mybe->port);
            return true; // Trigger connection failure
        } else if (mysql_thread___session_track_variables == OPTIONAL) {
            proxy_info("Backend %s:%d does not support session tracking, continuing without it",
                      mybe->host, mybe->port);
            return false; // Skip session tracking setup
        }
        return false;
    }

    // Check CLIENT_DEPRECATE_EOF requirement
    if ((mybe->server_myds->myconn->mysql->server_capabilities & CLIENT_DEPRECATE_EOF) == 0) {
        if (mysql_thread___session_track_variables == ENFORCED) {
            proxy_error("Backend %s:%d does not support CLIENT_DEPRECATE_EOF required for session tracking",
                       mybe->host, mybe->port);
            return true; // Trigger connection failure
        } else if (mysql_thread___session_track_variables == OPTIONAL) {
            proxy_info("Backend %s:%d does not support CLIENT_DEPRECATE_EOF, session tracking will be limited",
                      mybe->host, mybe->port);
            // Could proceed with limited functionality or skip entirely
        }
    }

    // Proceed with session tracking setup...
}
```


## Summary

This PR implements valuable functionality that addresses real production needs. However, the current implementation has critical gaps that could cause connection failures in production environments.

**Critical Issues Summary**:
1. **Protocol Capability Gap**: Missing `CLIENT_SESSION_TRACK` detection will cause failures on older MySQL/MariaDB versions
2. **Dependency Gap**: Missing `CLIENT_DEPRECATE_EOF` handling

**Recommendation**:
- **Address critical issues first** before merge consideration
- **Implement enhanced configuration modes** for production robustness
- **Add comprehensive testing** across different MySQL versions and configurations
- **Consider Phase 2 enhancements** (frontend propagation) as follow-up PRs

The architectural foundation is solid and the approach is sound - we just need to harden the implementation for production use across diverse backend environments.

wazir-ahmed committed:
Enable session tracking based on server capabilities

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -1908,10 +1908,12 @@ bool MySQL_Session::handler_again___verify_backend_session_track_gtids() {
-		return ret; // exit immediately
+	if ((mybe->server_myds->myconn->mysql->server_capabilities & CLIENT_SESSION_TRACKING) == 0
+		|| (mybe->server_myds->myconn->mysql->server_capabilities & CLIENT_DEPRECATE_EOF) == 0
+		|| mysql_thread___enable_server_deprecate_eof == false) {
```
The variable may have been changed after connections were established, thus this check is unnecessary.
Also, we would implement a logic that if session_track_variables is enabled, enable_server_deprecate_eof is ignored and CLIENT_DEPRECATE_EOF is always enabled for backend connections

renecannao commented on lib/mysql_connection.cpp:
```diff
@@ -866,13 +868,18 @@ void MySQL_Connection::connect_start_SetClientFlag(unsigned long& client_flags)
 
+	// override 'mysql-enable_server_deprecate_eof' behavior if 'session_track_variables' is set to 'ENFORCED'
+	if (mysql_thread___session_track_variables == session_track_variables::ENFORCED) {
+		mysql->options.client_flag |= CLIENT_DEPRECATE_EOF;
```
Yeah, that is what I mean in my previous comment:
"f session_track_variables is enabled, enable_server_deprecate_eof is ignored and CLIENT_DEPRECATE_EOF is always enabled for backend connections"

renecannao commented on lib/mysql_connection.cpp:
```diff
@@ -866,13 +868,18 @@ void MySQL_Connection::connect_start_SetClientFlag(unsigned long& client_flags)
 
+	// override 'mysql-enable_server_deprecate_eof' behavior if 'session_track_variables' is set to 'ENFORCED'
+	if (mysql_thread___session_track_variables == session_track_variables::ENFORCED) {
+		mysql->options.client_flag |= CLIENT_DEPRECATE_EOF;
```
We should also add CLIENT_SESSION_TRACK

wazir-ahmed replied:
`CLIENT_SESSION_TRACK` is set by mariadb-client, by default. It is part of `CLIENT_CAPABILITIES` macro.

wazir-ahmed committed:
TAP: Add tests for session variable tracking

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed committed:
TAP: Add mysql56-single group

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

renecannao commented:
@coderabbitai full review

renecannao committed:
Merge branch 'v3.0' into session-track-system-variable

Signed-off-by: René Cannaò <rene@proxysql.com>

renecannao commented:
@coderabbitai full review

renecannao committed:
Merge branch 'v3.0' into session-track-system-variable

Signed-off-by: René Cannaò <rene@proxysql.com>

mirostauder commented:
retest this please

Commit:
fix syntax error in groups.json

Signed-off-by: Miro Stauder <miro@proxysql.com>

wazir-ahmed committed:
TAP: Add 'no-repl' flag for mysql56-single group

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed committed:
TAP: Restructure tests for session variable tracking

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed committed:
TAP: Update mysql_user for session variable tracking tests

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

mirostauder commented:
retest this please

renecannao committed:
Merge branch 'v3.0' into session-track-system-variable

Signed-off-by: René Cannaò <rene@proxysql.com>

wazir-ahmed committed:
Merge branch 'v3.0' into session-track-system-variable

wazir-ahmed committed:
CI: Add `mysql56-single` test infra

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed committed:
chore(tap): Fix lint errors in groups.json

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed committed:
fix: preserve EOF capability in enforced mode

- Keep `CLIENT_DEPRECATE_EOF` enabled during frontend handshake when
`session_track_variables=ENFORCED`.
- Avoid unsafe mysql_error() calls in session-track TAP tests.

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed commented:
@coderabbitai Do a full review of this PR. Report if you find any new issues. Don't repeat previous review comments.

wazir-ahmed committed:
fix: skip session variable handling when tracking is disabled

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed committed:
fix: make server_backoff_time atomic

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

renecannao committed:
MySrvC: rename server_backoff_time to session_track_backoff_until, gate hot-path read

Context
-------
PR #5166 introduced a per-'MySrvC' atomic timestamp that prevents the server
from being chosen for new connections for a 30-second window. The timestamp
is written in exactly one place, 'MySQL_Session::handle_session_track_capabilities()',
and only when 'mysql-session_track_variables' is in ENFORCED mode and the
backend is missing 'CLIENT_SESSION_TRACKING' or 'CLIENT_DEPRECATE_EOF'.

The field was originally named 'server_backoff_time' and its only semantics
were that ENFORCED-mode capability mismatches temporarily exclude the server
from selection. Two issues with the original shape:

  1. The name does not hint that the scope is session-tracking capability
     enforcement. A reader of 'MyHGC::get_random_MySrvC' or
     'MySQL_Thread::get_MyConn_local' had no way to tell that this field is
     tied to one specific feature and not, say, a generic health backoff.

  2. The atomic load was unconditional in both hot selection loops. Even
     though the atomic is relaxed (effectively a 'mov' on x86), per-iteration
     atomic reads inhibit hoisting and CSE that the compiler can otherwise
     apply to thread-local integer reads. 'get_random_MySrvC' runs the loop
     over every server in the hostgroup on every backend-connection
     acquisition, so this is the sort of place where small per-iteration
     costs accumulate.

Changes
-------
* Rename the field:
    MySrvC::server_backoff_time  ->  MySrvC::session_track_backoff_until
  Naming it after the feature makes it obvious that the field is specific
  to session-tracking enforcement and should not be overloaded for unrelated
  purposes (e.g. connect errors, monitor failures). If a future change needs
  a generic backoff mechanism, a new field -- or a rename at that point --
  keeps intent clear rather than piggy-backing on this one silently.

* Gate the atomic load on mode:
    In both 'MyHGC::get_random_MySrvC()' (lib/MyHGC.cpp) and
    'MySQL_Thread::get_MyConn_local()' (lib/MySQL_Thread.cpp), the relaxed
    load is skipped entirely when 'mysql_thread___session_track_variables'
    is not ENFORCED. In 'get_MyConn_local' the check is computed once before
    the loop; in 'get_random_MySrvC' the check is inside the loop but is a
    thread-local int read, which is free.

    Correctness note: switching out of ENFORCED at runtime makes any stale
    'session_track_backoff_until' values in the pool immediately ignored,
    which is the desired behavior -- servers rejected while ENFORCED was
    active become eligible again the moment the operator relaxes the mode,
    matching how ProxySQL handles other runtime-variable-dependent state.
    No explicit reset is required.

* Updated the field documentation in include/MySQL_HostGroups_Manager.h
  with the monotonic-timestamp unit, the list of sites that read/write it,
  the explicit scope (ENFORCED mode only), and the reset policy (no explicit
  reset; deadline ages out by comparison against curtime).

No functional change for DISABLED / OPTIONAL configurations (the default for
all installations). ENFORCED mode behaviour is unchanged.

renecannao committed:
session_track_variables: add observability for misconfiguration and runtime backoff

Context
-------
PR #5166 introduced a three-mode 'mysql-session_track_variables' variable
(DISABLED / OPTIONAL / ENFORCED) whose behavior silently depends on two
other MySQL capability toggles and on the capabilities of each backend:

  * In OPTIONAL mode the feature is a no-op when
    'mysql-enable_server_deprecate_eof=false' or when the backend does not
    advertise CLIENT_DEPRECATE_EOF / CLIENT_SESSION_TRACKING. The operator
    sees the variable set to 1, but 'PROXYSQL INTERNAL SESSION' shows no
    tracked variables and there is no indication why.

  * In ENFORCED mode ProxySQL forces CLIENT_DEPRECATE_EOF on both the client
    and server sides of the protocol even if the operator explicitly set
    'mysql-enable_client_deprecate_eof=false' / 'mysql-enable_server_deprecate_eof=false'.
    This is a protocol-visible behavior change for clients that did not
    negotiate CLIENT_DEPRECATE_EOF, and it was previously silent.

  * In ENFORCED mode backends that cannot honor tracking (missing
    CLIENT_SESSION_TRACKING or CLIENT_DEPRECATE_EOF) are placed in a
    30-second selection backoff via 'MySrvC::session_track_backoff_until'.
    The server is NOT shunned and no connect-error counter is incremented,
    so in a mixed-version hostgroup an incompatible backend sits idle
    indefinitely with no log trail.

Changes
-------
* lib/Admin_FlushVariables.cpp: add a cross-variable validator at the tail
  of 'ProxySQL_Admin::flush_mysql_variables___database_to_runtime()', right
  next to the existing 'mysql-use_tcp_keepalive' warning. It runs once per
  'LOAD MYSQL VARIABLES TO RUNTIME' and only when the feature is actually
  requested (mode != DISABLED). Emitted warnings:
    - ENFORCED + !enable_client_deprecate_eof -> "will force CLIENT_DEPRECATE_EOF
      on frontend connections; older clients may be affected"
    - ENFORCED + !enable_server_deprecate_eof -> equivalent for backends
    - OPTIONAL + !enable_server_deprecate_eof -> "has no effect while
      mysql-enable_server_deprecate_eof=false; backend session tracking will
      be skipped"
  The checks use the global 'GloMTH->get_variable_int()' view (same pattern
  as the TCP keepalive warning above it), and deliberately do not name any
  backend -- this is a global-configuration warning; per-backend capability
  mismatches are handled separately below.

* lib/MySQL_Session.cpp: in 'handle_session_track_capabilities()', at the
  point where the code stores a new 'session_track_backoff_until' deadline
  for an incompatible backend, add a 'proxy_warning' naming the backend's
  address:port and the reason. Because this path only runs on new
  connection establishment -- and is itself gated by the 30-second backoff
  on the selection path -- the warning is naturally rate-limited to at most
  once per 30 seconds per backend. In a mixed fleet that rate is just
  enough for an operator to notice the incompatible server without drowning
  the log.

Behaviour
---------
* DISABLED mode (the default): no new log output; validator is skipped
  entirely. The hot selection path was already skipping the atomic read in
  the previous commit, so DISABLED remains free of new work.

* OPTIONAL / ENFORCED: at most three warnings at 'LOAD MYSQL VARIABLES TO
  RUNTIME' time (cross-variable validator) plus one periodic warning per
  incompatible backend (ENFORCED path only). No per-query log output.

renecannao committed:
TAP: parameterize session_track_variables helpers; add time_zone and sql_mode coverage

Context
-------
The four session_track_variables TAP tests previously hard-coded a single
variable, 'innodb_lock_wait_timeout', for their round-trip assertion. That
only exercised the integer path in
'MySQL_Session::handler_rc0_Process_Variables', leaving the string-value
path (and the character-set special case elsewhere in the same function)
untested. The stored-procedure helper was also named generically
('set_session_variable') while its body was pinned to InnoDB's lock wait
timeout -- confusing for readers.

Changes
-------
* test/tap/tests/session_track_variables.h:
    - Introduce 'struct tracked_var_spec { name, set_expr; }' so the helpers
      can exercise arbitrary session variables from a single
      spec-driven code path.
    - 'set_session_variable()' now takes a spec, builds a per-variable
      stored procedure name ('test.set_session_track_<name>') so multiple
      specs can coexist without stepping on each other, and guarantees
      that the CALL is the last statement on the connection before the
      tracker is read (any query in between would reset the session
      tracker state).
    - Add 'drop_session_variable_proc()' for optional teardown. Kept
      separate from the set path so tests that need to preserve the CALL
      as the last statement can skip cleanup.
    - Switch the whole value surface from 'int' to 'std::string':
      'extract_sess_var_mysql_pkt', 'select_sess_var_value', and
      'extract_sess_var_proxy_internal' all return strings, which is how
      session tracking delivers values on the wire and matches both
      integer and string variables without encoding assumptions.
    - Provide a default set, 'session_track_default_vars', with three
      representative cases:
        * 'innodb_lock_wait_timeout' (integer; original coverage)
        * 'time_zone' (string, portable small value-space)
        * 'sql_mode' (string, representative of production usage)
      All three 'set_expr' values are pure server-side SQL (no client-side
      randomization), so a CALL alone produces a fresh value per run.

* test/tap/tests/mysql-session_track_variables_optional-t.cpp,
  test/tap/tests/mysql-session_track_variables_enforced-t.cpp,
  test/tap/tests/mysql-session_track_variables_ff-t.cpp,
  test/tap/tests/mysql-session_track_variables_ff_mysql56-t.cpp:
    - Iterate over 'session_track_default_vars' and emit one TAP 'ok' per
      (spec[, mode]) combination.
    - Adjust plans accordingly. The ENFORCED-on-5.6 case still emits a
      single 'ok' because the backend is rejected before any per-variable
      tracking is attempted, so multiplying by the spec count there would
      be misleading.
    - Compare values as strings using the new string-returning helpers.

The broadened coverage does not change what 'handler_rc0_Process_Variables'
or 'handle_session_track_capabilities' do; it only exercises more of their
existing behavior, notably the non-integer value path and the
'proxysql_find_charset_*' branches that activate for character-set
variables (not exercised by these three specs but now reachable through
the same harness for follow-up tests).

renecannao committed:
ci: add CI-mysql56-single-g1 caller for the mysql56-single-g1 TAP group

Context
-------
PR #5166 introduced a new TAP group 'mysql56-single-g1' backed by
'test/infra/infra-dbdeployer-mysql56-single/' (a dbdeployer-based MySQL 5.6
fixture). Four new tests register against this group in
'test/tap/groups/groups.json':

  - mysql-session_track_variables_enforced-t
  - mysql-session_track_variables_ff_mysql56-t
  - mysql-session_track_variables_optional-t

Without a dedicated CI workflow these tests are never exercised
automatically: the generic 'CI-taptests-groups' plumbing hardcodes
'infradb: [ mysql57 ]' in its matrix, and no existing 'CI-mysql*' caller
routes to the 'mysql56-single-g1' group.

Changes
-------
This commit adds the v3.0-branch half of the standard caller/reusable
pair, following the architecture documented in 'doc/GH-Actions/README.md'
(section "The dedicated-reusable pattern"). The file is modelled on
'CI-mysql84-g1.yml' verbatim -- only the workflow name and the
'uses:' target change. Like every other 'CI-*.yml' caller on v3.0:

  - Triggers on 'workflow_run' after 'CI-trigger' completes successfully
    (so the build/test caches are guaranteed populated) plus manual
    'workflow_dispatch'.
  - Uses 'concurrency.cancel-in-progress: true' to cancel any in-flight
    run when a new push lands on the same branch.
  - Delegates all real work to 'ci-mysql56-single-g1.yml' on the
    'GH-Actions' branch.

The reusable half of the pair -- 'ci-mysql56-single-g1.yml' on the
'GH-Actions' branch -- is added in a separate commit on that branch.
Until both halves are present this caller will fail with
"workflow not found". That is the expected landing order per the
project's CI workflow: the reusable ships first on 'GH-Actions', then
the caller ships on v3.0, matching how 'CI-mysql84-g1.yml' was
introduced (see 'doc/GH-Actions/README.md' section "Adding a new test
group end-to-end").

renecannao commented:
@wazir-ahmed quick question on the test infra layout. This PR ships two parallel MySQL 5.6 fixtures:

- `test/infra/infra-dbdeployer-mysql56-single/` — dbdeployer-based, wired up as the default by `test/tap/groups/mysql56-single/env.sh`, `infras.lst`, and the `run-tests-isolated.bash` case statement.
- `test/infra/infra-mysql56-single/` — a Docker-Compose / Dockerfile variant. Grepping the tree, nothing outside its own directory references it.

We're moving away from Docker-Compose for single-instance MySQL fixtures in favor of dbdeployer. If the Docker variant is not serving a purpose I'm missing, it looks like a candidate for removal in this PR (or a follow-up) — can you confirm whether it's safe to drop? If it's kept as an intentional alternative, a note in `test/infra/README.md` explaining when to use each would help future readers.

wazir-ahmed committed:
chore(CI): Remove regular mysql56-single infra

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed commented:
@renecannao I have removed the dockerized infra. We now have only one `mysql56-single` infra which is based on `dbdeployer`.
